### PR TITLE
Switched to the new alpaca-py library

### PR DESCRIPTION
**Pull request by grzesir**

### CODE REVIEW

**Review by davidlatte — APPROVED** [review]
Looks pretty good, I don't see any required hard changes.  Have a few cleanup comments.

File: requirements.txt [hunk 1]
```diff
@@ -1,6 +1,5 @@
 polygon-api-client
 aiohttp==3.8.1  # 4.0.0a1 is installed by default but aiohttp==3.8.1 is required by {'alpaca-trade-api'}
```
[the remark this answers is not in the document]
Comment (grzesir) [reply requirements.txt:2]: fixed. ideally this is synced with setup.py somehow. always forget to update this file

File: requirements.txt [hunk 2]
```diff
@@ -26,4 +25,5 @@ python-dotenv  # Secret Storage
 ccxt==3.0.61
 termcolor
 jsonpickle
-apscheduler==3.10.1
+apscheduler==3.10.4
+alpaca-py<=0.10.0
```
[the remark this answers is not in the document]
Comment (grzesir) [reply requirements.txt:29]: great idea! i did this with a bunch of other libraries too

File: lumibot/brokers/alpaca.py [hunk 3]
```diff
@@ -217,29 +218,29 @@ def _get_balances_at_broker(self, quote_asset):
...
         response = self.api.get_account()
-        total_cash_value = float(response._raw["cash"])
-        gross_positions_value = float(response._raw["long_market_value"]) - float(
-            response._raw["short_market_value"]
+        total_cash_value = float(response.cash)
+        gross_positions_value = float(response.long_market_value) - float(
+            response.short_market_value
         )
-        net_liquidation_value = float(response._raw["portfolio_value"])
+        net_liquidation_value = float(response.portfolio_value)
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/brokers/alpaca.py:225]: i dont think so. that could cause issues in crypto, where some have 32+ decimal places

File: lumibot/brokers/alpaca.py [hunk 4]
```diff
@@ -266,10 +267,18 @@ def map_asset_type(self, type):
     def _parse_broker_order(self, response, strategy_name, strategy_object=None):
         """parse a broker order representation
         to an order object"""
+
+        # If the symbol includes a slash, then it is a crypto order and only the first part of
+        # the symbol is the real symbol
+        if "/" in response.symbol:
+            symbol = response.symbol.split("/")[0]
+        else:
+            symbol = response.symbol
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/brokers/alpaca.py:276]: honestly, I'm not a big fan of the ternary operator. to me, it makes it harder to read. would be open to debate this though

File: lumibot/brokers/alpaca.py [hunk 5]
```diff
@@ -352,14 +378,18 @@ def _submit_order(self, order):
...
-        else:
-            trade_symbol = order.asset.symbol
-
         try:
-            qty = str(order.quantity)
-            response = self.api.submit_order(trade_symbol, qty, order.side, **kwargs)
+            # Create our own OrderData class to pass to the API because this is easier to work with
+            # than the ones Alpaca provides, and because the new classes are missing bracket orders
+            class OrderData:
+                def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/brokers/alpaca.py:385]: good idea! done

File: lumibot/brokers/alpaca.py [hunk 6]
```diff
@@ -472,7 +490,7 @@ async def _trade_update(trade_update):
                 )
 
                 return True
-            except:
+            except Exception:
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/brokers/alpaca.py:493]: done

File: lumibot/data_sources/alpaca_data.py [hunk 7]
```diff
@@ -48,68 +48,77 @@ def __init__(self, config, max_workers=20, chunk_size=100, **kwargs):
...
         else:
             raise ValueError("API_SECRET not found in config")
 
-        if type(config) == dict and "ENDPOINT" in config:
-            self.endpoint = config["ENDPOINT"]
-        elif hasattr(config, "ENDPOINT"):
-            self.endpoint = URL(config.ENDPOINT)
+        # If an ENDPOINT is provided, warn the user that it is not used anymore
+        # Instead they should use the "PAPER" parameter, which is boolean
+        if isinstance(config, dict) and "ENDPOINT" in config:
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/data_sources/alpaca_data.py:67]: yea im glad they made this change, overall their new library seems to have better design choices

File: lumibot/data_sources/alpaca_data.py [hunk 8]
```diff
@@ -130,47 +139,54 @@ def get_barset_from_api(
             limit = 1000
 
         if end is None:
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/data_sources/alpaca_data.py:141]: what do you mean? "if end" is the opposite, no?

File: lumibot/entities/order.py [hunk 9]
```diff
@@ -432,6 +432,18 @@ def quantity(self, value):
             quantity, "Order quantity must be a positive Decimal"
         )
 
+    def __hash__(self):
+        return hash(self.identifier)
+
+    # Compares two order objects to see if they are the same.
+    def __eq__(self, other):
```
[the remark this answers is not in the document]
Comment (grzesir) [reply lumibot/entities/order.py:439]: @davidlatte fixed. also really cool that it wrote the code, all I had to do was copy/paste it. this thing is pretty impressive

**Pull request merged**